### ISSUE TITLE
Possible error following multiple relative symlinks from /etc

### ISSUE THREAD

**Issue by kyren** [NixOS/nixpkgs#56278]
## Issue description

Okay, I'm not sure whether this is an issue with nixpkgs or nix or what precisely, but sometime in the last few days my nixos desktop has started to error when things look up `<nixpkgs>` from `NIX_PATH`.

For example:

```
 % nix-shell -p hello -vvvv
evaluating file '/nix/store/h0fcdw5bgg9yd5mr0v5dip4xi0vrp5m6-nix-2.2/share/nix/corepkgs/derivation.nix'
resolved search path element '/home/catherine/.nix-defexpr/channels' to '/home/catherine/.nix-defexpr/channels'
resolved search path element '/etc/nixpkgs' to '/etc/nixpkgs'
error: getting status of '/xp7gm5nq3r2m16cwhfk71kwqj4kgyw4r-kqqvh23bi5kfxm7z3mj30cn67myn9s72-19eedaf867da3155eec62721e0c8a02895aed74b.tar.gz': No such file or directory
```

Currently I have nixpkgs copied to /etc/ and my `NIX_PATH` set as `nixpkgs=/etc/nixpkgs` like this:

```
{ config, pkgs, ... }:
{
  nix = {
    nixPath = [ "nixpkgs=/etc/nixpkgs" ];
  };

  environment.etc = [
    {
      source = pkgs.path;
      target = "nixpkgs";
    }
  ];
}
```

Up until a few days ago this was working fine for me, and nothing has changed other than updating nixpkgs.

After talking about the issue on IRC, it was suggested I strace a nix-shell invocation, which I did like this:

```
strace -o logfile -ff nix-shell -p hello -vvvv
```

and got [this output](https://bepasty.kyju.org/zg2mmXpk/+inline?).  The salient lines here seem to be around:

```
readlink("/etc/static/nixpkgs", "../../xp7gm5nq3r2m16cwhfk71kwqj4"..., 1024) = 119
```

which makes it seem like something is not reading relative symlinks correctly when following multiple symlinks?

### Steps to reproduce

It should be reproducible by adding the above NixOS module to a new install and running `nix-shell -p hello`.

## Technical details

> Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the results.

WELL I can't exactly do that right now 😉, but the git sha I'm using currently is 19eedaf867da3155eec62721e0c8a02895aed74b.

Edit: Okay, this IS NixOS, of course can run it... I just have to try harder:

```
 % NIX_PATH=nixpkgs=https://github.com/NixOS/nixpkgs-channels/archive/19eedaf867da3155eec62721e0c8a02895aed74b.tar.gz nix-shell -p nix-info --run 'nix-info -m'
unpacking 'https://github.com/NixOS/nixpkgs-channels/archive/19eedaf867da3155eec62721e0c8a02895aed74b.tar.gz'...
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/0anvhqarc54r0vvv6v1hx6yg6g97c2sf-bash-interactive-4.4-p23-dev
copying path '/nix/store/0anvhqarc54r0vvv6v1hx6yg6g97c2sf-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.20.10, NixOS, 19.03pre-git (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `""`
 - channels(catherine): `""`
 - nixpkgs: `/nix/store/kqqvh23bi5kfxm7z3mj30cn67myn9s72-19eedaf867da3155eec62721e0c8a02895aed74b.tar.gz`
```

Obviously that's setting NIX_PATH manually, but that's what I am using to deploy my desktop OS now so it should match.

Edit 2:

I've switched to another method of setting `NIX_PATH` that works around this bug, and after upgrading I now have a dead symlink in `etc`:

```
 % ls -l /etc/nixpkgs
lrwxrwxrwx 1 root root 14 Feb 23 23:36 /etc/nixpkgs -> static/nixpkgs
```

It is my understanding that this should have been cleaned up on the next `nixos-rebuild`?

It was suggested on IRC (again by `clever`, who's been helping me through this) that I add this to the bug report since it may be related to the bug.

**Comment by cleverca22** [comment]
https://github.com/NixOS/nixpkgs/pull/54980 may be related?

**Comment by peterhoeg** [comment]
I'm seeing some possibly related behaviour here when doing ```nixos-rebuild``` as well on latest unstable:

```
duplicate entry X11/xkb -> /nix/store/kv54wh70md6cv30f9dj91g600lkg01md-xkeyboard-config-2.24/etc/X11/xkb
mismatched duplicate entry ../../../kv54wh70md6cv30f9dj91g600lkg01md-xkeyboard-config-2.24/share/X11/xkb <-> /nix/store/kv54wh70md6cv30f9dj91g600lkg01md-xkeyboard-config-2.24/etc/X11/xkb
builder for '/nix/store/nx29bx0mbp0bmp0g8pi834a3x43rds2h-etc.drv' failed with exit code 1
```

It's not a duplicate entry as those are the same.

**Comment by danbst** [comment]
just reverted https://github.com/NixOS/nixpkgs/pull/54980. Sorry for not doing it sooner

**Comment by peterhoeg** [comment]
As it turns out, my case was caused by the combination of this and ```services.xserver.exportConfiguration = true;``` - changing this to false made things work again.

**Comment by kyren** [comment]
This seems to be resolved now that #54980 is reverted 👍
